### ISSUE TITLE
[Storage] Rate Limit

### ISSUE THREAD

**Issue by devmanhinton** [Azure/azure-cli#1118]
`az storage blob upload: error: Azure Error: TooManyRequests`

Was using JavaScript (Node) lib before and never saw this issue. 

Need to copy ~ `1000` blobs between containers.
Tried using `copy`. Tried using `download` and then `upload`. 

You all doing anything weird or have some default on this

**Comment by mayurid** [comment]
@tjprescott to take a look

**Comment by devmanhinton** [comment]
Thanks! Seems super strange. Getting capped after like 50 requests which seems absurd. Even tried adding in a sleep every 20 requests

**Comment by tjprescott** [comment]
Any rate limiting would be imposed by the service, not by the CLI, as both CLIs (Python and Node) simply call into their respective SDKs. @rambo-msft can you provide guidance on this?

**Comment by devmanhinton** [comment]
Okay hmmm. I'll try to dig into the `scala` lib one of our engineers built. It can download all 1300 blobs in about 5 or something minutes without any rate limiting.

Also have not had time to investigate but as an FYI I think installing this CLI breaks the Node one.

1. Setup NODE CLI
1. Use CLI w/ credentials
1. Install this CLI
1. Use `az configure`
1. Use new CLI

--> 

Node CLI breaks w/
```
SyntaxError: Unexpected token
    at Object.parse (native)
    at loadProfileFromFile (/usr/local/azure/lib/util/profile/profile.js:287:46)
    at loadDefaultProfile (/usr/local/azure/lib/util/profile/profile.js:277:12)
    at load (/usr/local/azure/lib/util/profile/profile.js:266:12)
    at Object.<anonymous> (/usr/local/azure/lib/util/profile/profile.js:339:22)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:313:12)
    at Module.require (module.js:366:17)
```

rm -rf  `~/.azure/` fixes the error. 

It is possible I did something in the interim but doubt it

**Comment by tjprescott** [comment]
> Also have not had time to investigate but as an FYI I think installing this CLI breaks the Node one.

@yugangw-msft this sounds like it may be related to cred sharing. Can you take a look? Either way, we should put this in a separate issue.

**Comment by yugangw-msft** [comment]
Please open a separate issue, what is the node cli's version? Please run `azure -v` to verify you are using the newer one (0.10.5+)

**Comment by tjprescott** [comment]
@troydai with the work you are doing on revamping the storage commands, you would be most likely to run into this issue.

**Comment by johanste** [comment]
There are a couple of root causes for this issue. The first thing is to determine if the command is running into the management operation API quota limit (12000 calls/hr). The second thing is to see if it is related to the IOPS for the storage account (very unlikely). 

Assuming that we are not doing an excessive amount of management calls/command invocation, there is little we can do in the CLI other than suggesting that the caller provides the keys instead of retrieving it for each invocation for bulk operations.

**Comment by troydai** [comment]
So far I can't repo this issue. I do omit the account key to force more Azure API requests. When copy a small file repeatedly in single thread the average upload takes 3.5 seconds (for 4 B file). The script manages to keep uploading more than 100 files without any error. However, I do think the performance is problematic for such a small file. 
I also increase the concurrency by starting 20 threads to upload files in parallel. I couldn't hit the rate either.

@devmanhinton does this issue sill bother you? Can you share the script in which you can repo this issue? What's the configuration of your storage account?

Thanks,
Troy

**Comment by troydai** [comment]
@devmanhinton In my test, I start 20 threads to aggressively upload blob. I can hit the limit quickly if I don't give the account key. In this case, CLI will have to send request to management API for storage account credential. 

The bottleneck happens to this request:
```
https://management.azure.com:443 "GET /subscriptions/00977cdb-163f-435f-9c32-39ec8ae61f4d/providers/Microsoft.Storage/storageAccounts?api-version=2016-01-01 HTTP/1.1" 429 80
```

Once I set the key on the command line (equivalent to set in the environment variable), I'm not able to hit the limit again.

Could you give it a try?

**Comment by troydai** [comment]
For upload command where account key is not specified, two additional management requests are required:
1. GET /subscriptions/<id>/providers/Microsoft.Storage/storageAccounts
2. POST /subscriptions/<id>/resourceGroups/<name>/providers/Microsoft.Storage/storageAccounts/<account>/listKeys

They're executed without cache for every request. The API rate limit will be hit fast in this situation.

**Comment by troydai** [comment]
@johanste I close this since there is nothing we can do if the key is not given. Let me know or reopen if you think there are other approaches for this problem.

**Comment by tjprescott** [comment]
@troydai for the batch upload command, the key should only be requested once. So the batch upload command should also alleviate @devmanhinton's issue

**Comment by troydai** [comment]
@tjprescott good point. You're right. Since the client is generated once. 

@devmanhinton we introduced batch-copy command a few weeks back. It will not only help to solve the api rate limit, but also support copy blob between containers. Please give it a try.

**Comment by devmanhinton** [comment]
@troydai Awesome. Thanks! Sorry for lack of response. Things got crazy and we want with our own custom solution and then using Azure Storage Explorer (with drop and drag ability). We are moving towards some automated deploy scripts moving around resources and will circle back on this when we get to that point. Thank you for the followup!

**Comment by troydai** [comment]
No problem, @devmanhinton . Please feel free to re-open this issue if existing solutions don't fit your scenario. Thanks.

**Comment by devmanhinton** [comment]
Will do! Thank you!

**Comment by devmanhinton** [comment]
@troydai trying now and running into errors with  `az storage blob copy start-batch ...`

#### Bad Error message?
1. Forget to have pattern
```
az storage blob copy start-batch: error: incorrect usage: --recursive | --pattern PATTERN
```
1. Add pattern `--pattern '*.csv'`
1. Errors
```
az: error: unrecognized arguments: --pattern *.csv
```

#### Does not Work?

1. Test copy from redacted-f --> redacted-t in redacted-account

```
az storage blob copy start-batch --recursive --destination-container redacted-t  --source-account redacted-account --source-container redacted-f
```

1. Errors
```
The specified resource does not exist.
```
1. `--verbose` yields not more info
1. listing both containers shows they exist and have data
```
az storage blob list -c redacted-f --account-name redacted-account
az storage blob list -c redacted-t --account-name redacted-account
```

FYI -- I installed using the steps in the readme `curl -L https://aka.ms/InstallAzureCli | bash`

Any idea?
